nn/conformer.py

```diff
@@ -0,0 +1,204 @@
+    for filter_size, channel_size in zip(filter_sizes, channel_sizes):
+      self.conv_layers.append(
+        nn.Conv(filter_size=filter_size, n_out=channel_size, padding=padding))
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    x = nn.split_dims(inp, axis='F', dims=(-1, 1))
+    for i, conv_layer in enumerate(self.conv_layers):
+      x = conv_layer(x)
+      x = self.activation(x)
+      if self.pool_sizes and i < len(self.pool_sizes):
+        x = nn.pool(x, pool_size=self.pool_sizes[i], padding='same', mode='max')
+      if self.dropout:
+        x = nn.dropout(x, dropout=self.dropout)
+    out = nn.merge_dims(x, axes='static')
+    return out
+
+
+class ConformerEncoderLayer(nn.Module):
+  """
+  Represents a conformer block
+  """
+
+  def __init__(
+        self, conv_kernel_size: int = 32, activation_ff: Callable[[nn.LayerRef], nn.LayerRef] = nn.swish,
+        dim_ff: int = 2048, dropout: float = 0.1, att_dropout: float = 0.1, out_dim: int = 512, num_heads: int = 8,
+        batch_norm_opts: Optional[Dict[str, Any]] = None):
+    """
+    :param conv_kernel_size:
+    :param activation_ff:
+    :param dim_ff:
+    :param dropout:
+    :param att_dropout:
+    :param out_dim:
+    :param num_heads:
+    :param batch_norm_opts:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.out_dim = out_dim
+
+    self.ffn1 = _PositionwiseFeedForward(
+      out_dim=out_dim, dim_ff=dim_ff, dropout=dropout, activation=activation_ff)
+
+    self.ffn2 = _PositionwiseFeedForward(
+      out_dim=out_dim, dim_ff=dim_ff, dropout=dropout, activation=activation_ff)
+
+    self.conv_block = _ConformerConvBlock(
+      out_dim=out_dim, kernel_size=conv_kernel_size, batch_norm_opts=batch_norm_opts)
+
+    self.self_att = MultiheadAttention(enc_key_dim, num_heads, dropout=att_dropout)  # TODO: to be implemented
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    # FFN
+    x_ffn1_ln = nn.layer_norm(inp)
+    x_ffn1 = self.ffn1(x_ffn1_ln)
+    x_ffn1_out = 0.5 * nn.dropout(x_ffn1, dropout=self.dropout) + inp
+
+    # MHSA
+    x_mhsa_ln = nn.layer_norm(x_ffn1_out)
+    x_mhsa = self.self_att(x_mhsa_ln)
+    x_mhsa_out = x_mhsa + x_ffn1_out
+
+    # Conv
+    x_conv_ln = nn.layer_norm(x_mhsa_out)
+    x_conv = self.conv_block(x_conv_ln)
+    x_conv_out = nn.dropout(x_conv, dropout=self.dropout) + x_mhsa_out
+
+    # FFN
+    x_ffn2_ln = nn.layer_norm(x_conv_out)
+    x_ffn2 = self.ffn2(x_ffn2_ln)
+    x_ffn2_out = 0.5 * nn.dropout(x_ffn2, dropout=self.dropout) + x_conv_out
+
+    # last LN layer
+    return nn.layer_norm(x_ffn2_out)
+
+
+class ConformerEncoder(nn.Module):
+  """
+  Represents Conformer encoder architecture
+  """
+
+  def __init__(self, encoder_layer: Union[ConformerEncoderLayer, Any], num_blocks: int):
+    """
+    :param encoder_layer:
+    :param num_blocks:
+    """
+    super().__init__()
+
+    self.dropout = getattr(encoder_layer, 'dropout')
+    out_dim = getattr(encoder_layer, 'out_dim')
```

Comment:
It should be consistent, also an attrib.
```suggestion
    self.out_dim = encoder_layer.out_dim
```